### PR TITLE
./lib Makefile rewrite, create libkhmer.{a,so} and install target

### PR DESCRIPTION
**Pull request by kdmurray91**
Hi all,

This rehashes the makefile in `./lib`, adding a few targets and cleaning up. (mentioned in #776, also ping #751)

Added:
- `libkhmer.so` shared object (how does this work on OSX? will it break due to shared objects being named X.dylib?)
- `libkhmer.a` static archive

removed:
- obsolete targets whose dependent source files don't exist any longer
- use wildcards for `*.o` targets where possible

changes:
- Use versioneer not python packaging for versioning. If haven't done `python setup.py something`, the old script spat out a sentence explaining this. Now, you don't need setup.py.
- add compiled files to gitignores

All of this stuff probably needs testing, this started as a hack to test a few things. There's probably also some more culling that could happen in the makefile.

Cheers,
K

### CODE REVIEW

**Comment by kdmurray91** [comment]
- [ ] Is it mergable?
- [ ] Did it pass the tests?
- [ ] If it introduces new functionality in scripts/ is it tested?
  Check for code coverage with `make clean diff-cover`
- [ ] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  `make cppcheck`, and `make doc` output. Use `make format` and manual
  fixing as needed.
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Is it documented in the ChangeLog?
  http://en.wikipedia.org/wiki/Changelog#Format
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?

**Comment by kdmurray91** [comment]
Whelp, merge conflict already :dizzy_face:. Will rebase and push.

**Comment by kdmurray91** [comment]
Hmm, are those failures my stuff up or Jenkins'?

**Comment by mr-c** [comment]
Yeah, this Makefile hasn't been used in about two years.

The following defines are no longer present in the codebase and can be removed: 

TRACE_STATE_CHANGES
TRACE_BUSYWAITS
TRACE_SPINLOCKS
TRACE_MEMCOPIES
TRACE_DATA

**Comment by kdmurray91** [comment]
retest this please

**Comment by kdmurray91** [comment]
@mr-c Does that mean that lib/trace_logger is also obsolete?

**Comment by mr-c** [comment]
@kdmurray91 Not quite yet :-)

**Comment by kdmurray91** [comment]
retest this please

**Comment by kdmurray91** [comment]
Can I delete the following files:

- `test-CacheManager.cc`: Functionality has been removed
- `test-StreamReader.cc`: Functionality has been removed

Additionally, `test-Colours.cc` doesn't compile.

#18

**Comment by kdmurray91** [comment]
I've added a simple build to the jenkins script, mostly to help me fix any bugs in the rewrite. Happy to take it out before merging

File: jenkins-build.sh [hunk 1]
```diff
@@ -68,4 +68,16 @@ fi
...
+pushd lib
+make clean
+make all
+make install PREFIX=../install_target
+popd
+test -d install_target/include
+test -f install_target/include/khmer.hh
+test -d install_target/lib
+test -f install_target/lib/libkhmer.a
+rm -rf install_target
```
Comment (mr-c) [inline jenkins-build.sh]: How about bundling this up as a target in the top-level Makefile?

File: lib/Makefile [hunk 2]
```diff
@@ -139,137 +93,124 @@ export CXX
...
-#DRV_PROGS=bittest # test-StreamReader test-CacheManager test-Parser test-HashTables
-DRV_PROGS+=#graphtest #consume_prof
-AUX_PROGS=#ht-diff
+BZIP2_OBJS=$(addprefix $(BZIP2_DIR)/, $(BZIP2_OBJS_BASE))
 
-CORE_OBJS= error.o khmer_config.o thread_id_map.o trace_logger.o perf_metrics.o kmer_hash.o
-PARSERS_OBJS= read_parsers.o
 
-all: $(ZLIB_OBJS) $(BZIP2_OBJS) $(CORE_OBJS) $(PARSERS_OBJS) hashtable.o hashbits.o subset.o counting.o hllcounter.o test aligner.o scoringmatrix.o node.o kmer.o
+#### Khmer proper below here ####
```
Comment (mr-c) [inline lib/Makefile]: s/Khmer/khmer/g

**Comment by mr-c** [comment]
+1 to deleting `test-CacheManager.cc` & `test-StreamReader.cc`

**Comment by mr-c** [comment]
And in general this LGTM; thanks! @luizirber & @camillescott may also have relevant feedback.

**Comment by kdmurray91** [comment]
jenkins retest this please

**Comment by kdmurray91** [comment]
is graphtest.cc ok to delete? seems obsolete too

**Comment by kdmurray91** [comment]
retest this please

**Comment by kdmurray91** [comment]
OK, so something about openmp is causing the OSX build slave to fail. I don't have any OSX machines handy, so can' t debug.

However, [this](https://stackoverflow.com/questions/20321988/error-enabling-openmp-ld-library-not-found-for-lgomp-and-clang-errors) suggests OpenMP isn't supported by clang on OSX "yet". Is this the issue at hand?

**Comment by luizirber** [comment]
@kdmurray91 on the Python side there is a [check][1] during installation. Apple's clang doesn't support OpenMP yet, despite clang 3.5 [changelog][2] saying it has support for parsing.

[1]: https://github.com/ged-lab/khmer/blob/9f82238d13ea4964063e1cd7998aab8645272612/setup.py#L52
[2]: http://llvm.org/releases/3.5.0/tools/clang/docs/ReleaseNotes.html#openmp-c-c-language-changes-in-clang

**Comment by kdmurray91** [comment]
aarrgh, jenkins retest this please you oaf

**Comment by camillescott** [comment]
my trend is spreading! <muahahaha>

**Comment by kdmurray91** [comment]
It's surprisingly therapeutic :)

**Comment by camillescott** [comment]
exactly! :)

**Comment by kdmurray91** [comment]
retest this please

**Comment by kdmurray91** [comment]
@camillescott @mr-c with regards to seqan, has the bundled version been patched at all? I.e., if there are seqan headers under /usr/include or /usr/local/include, will it work?

It does on my machine, but that could well be a fluke.

**Comment by mr-c** [comment]
@kdmurray91 Yep, see https://github.com/ged-lab/khmer/blob/master/third-party/seqan/ChangeLog

I wouldn't recommend using the system copy of Seqan at this time.

**Comment by kdmurray91** [comment]
I suspected so.

a638088 installs the seqan headers with the khmer headers.

However, this could clobber an existing seqan install.

We could install under `$PREFIX/include/khmer`, and make people set `-I $PREFIX/include/khmer` in `CFLAGS`.

Thoughts?

**Comment by kdmurray91** [comment]
Another option may be to change `seqan/` to `khmerseqan/` or something similar. But that seems a bit drastic.

**Comment by mr-c** [comment]
Why are we installing the seqan headers? The external interface is all khmer; our headers are enough to compile against and linking against a previously compiled dynamic library should be fine.

**Comment by kdmurray91** [comment]
Correct me if I'm wrong, but when I compile the following, `seqan/seqan.h` gets included:

```c++
#include  <counting.hh>

int main()
{
    khmer::CountingHash test(1,1);
    return 0;
}
```

At least, when I do something similar on a system with just the khmer headers installed, it fails due to not finding the seqan headers:

```
g++ -I $HOME/include -o tst tst.cc $HOME/lib/libkhmer.a
In file included from include/hashtable.hh:25:0,
                 from include/counting.hh:11,
                 from tst.cc:13:
include/read_parsers.hh:11:28: fatal error: seqan/sequence.h: No such file or directory
 #include <seqan/sequence.h>
                            ^
compilation terminated.
```

**Comment by mr-c** [comment]
Inclusion of the Seqan headers slows down compilation a lot. We can kill two birds with one stone here by using an opaque pointer to hide the implementation of `_stream` from the public header. This will also speed up compilation times for everyone. PR to your PR pending :-)

**Comment by mr-c** [comment]
See https://github.com/kdmurray91/khmer/pull/1 for my proposed fix. 

```
(env)mcrusoe@athyra:~/khmer/gl-master/install_target$ g++ -I include/ -o tst ../tst.cc lib/libkhmer.a
```

produces no error for me now

**Comment by kdmurray91** [comment]
Perfect, thanks very much @mr-c :)

**Comment by mr-c** [comment]
Jenkins, retest this please

**Comment by kdmurray91** [comment]
Jenkins, don't make us call Camille, retest this please!!

**Comment by kdmurray91** [comment]
Ah, I think jenkins is having trouble pulling down changes, as github's under a DDOS attack

**Comment by kdmurray91** [comment]
retest this please

**Comment by kdmurray91** [comment]
Well, closer. git fetch only failed on the OS X node that time. retest this please

**Comment by kdmurray91** [comment]
Ah, that's an *actual* failure. Hacky configure is hacky.

**Comment by kdmurray91** [comment]
image

File: Makefile [hunk 3]
```diff
@@ -184,6 +184,20 @@ lib:
...
+	mkdir -p install_target
+	cd lib && \
+	 $(MAKE) clean && \
+	 $(MAKE) all && \
+	 $(MAKE) install PREFIX=../install_target
+	test -d install_target/include
+	test -f install_target/include/khmer.hh
+	test -d install_target/lib
+	test -f install_target/lib/libkhmer.a
+	rm -rf install_target
```
Comment (mr-c) [inline Makefile]: Should your `tst.cc` be included here to catch issues similar to the Seqan header problem in the future?

File: lib/.check_openmp.cc [hunk 4]
```diff
@@ -0,0 +1,17 @@
+//
+// This file is part of khmer, http://github.com/ged-lab/khmer/, and is
+// Copyright (C) Michigan State University, 2009-2015. It is licensed under
```
Comment (mr-c) [inline lib/.check_openmp.cc]: Just Copyright 2015 please

File: lib/Makefile [hunk 5]
```diff
@@ -135,141 +89,167 @@ ifneq ($(shell uname), Linux)
 LIBS+= -pthread
 endif
 
-export CXX
-export CXXFLAGS
-export LIBS
 
-VERSION = $(shell ./get_version.py)
+HAVE_OPENMP=$(shell $(CXX) -fopenmp -o chkomp .check_openmp.cc 2>/dev/null && echo true || echo false; rm -f chkomp)
```
Comment (mr-c) [inline lib/Makefile]: Line wrapping would be great :-)

File: lib/khmer.pc.in [hunk 6]
```diff
@@ -0,0 +1,13 @@
+prefix=@prefix@
+exec_prefix=${prefix}
+libdir=${exec_prefix}/lib
+sharedlibdir=${libdir}
+includedir=${prefix}/include
+
+Name: khmer
+Description: khmer cpp library
```
Comment (mr-c) [inline lib/khmer.pc.in]: -> "The unsupported core C++ library from the khmer project"

+

`URL: http://khmer.readthedocs.org/`

**Comment by mr-c** [comment]
This generally looks good to me. Slap a completed checklist on this and I'll merge.

File: Makefile (hunk 3 above)
Comment (kdmurray91) [inline Makefile]: +1, done

File: lib/.check_openmp.cc (hunk 4 above)
Comment (kdmurray91) [inline lib/.check_openmp.cc]: Done

File: lib/Makefile (hunk 5 above)
Comment (kdmurray91) [inline lib/Makefile]: Ok, and all other lines >80 chars wrapped

File: lib/khmer.pc.in (hunk 6 above)
Comment (kdmurray91) [inline lib/khmer.pc.in]: done!

**Comment by kdmurray91** [comment]
- [ ] Is it mergeable?
- [ ] Did it pass the tests?
- [ ] If it introduces new functionality in scripts/ is it tested?
  Check for code coverage with `make clean diff-cover`
- [ ] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  `make cppcheck`, and `make doc` output. Use `make format` and manual
  fixing as needed.
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Is it documented in the ChangeLog?
  http://en.wikipedia.org/wiki/Changelog#Format
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?

**Comment by kdmurray91** [comment]
retest this please

**Comment by kdmurray91** [comment]
All good to go I think @mr-c

**Comment by mr-c** [comment]
Thanks!

**Pull request merged**